sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/AggregateFunction.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog.functions;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.DataType;
+
+/**
+ * Interface for a function that produces a result value by aggregating over multiple input rows.
+ * <p>
+ * The JVM type of result values produced by this function must be the type used by Spark's
+ * InternalRow API for the {@link DataType SQL data type} returned by {@link #resultType()}.
+ * <p>
+ * Most implementations should also implement {@link PartialAggregateFunction} so that Spark can
```

Comment:
Spark requires all aggregate functions to support partial merge, because the distributed aggregate operator in Spark requires it. As a result, even aggregate functions like `collect_list` have implemented the partial merging logic, though this doesn't make sense to them.

I don't think Spark can change this in the near future, and the current two UDAF APIs in Spark(`UserDefinedAggregateFunction` and `Aggregator`) also require implementing partial merge. I think we need to do the same here.

Reply:
@dongjoon-hyun, @sunchao, @viirya, @dbtsai, @wmoustafa, what do you think?

Sounds like we have 2 options:
1. Combine the partial interface with the aggregate interface so everything is required to implement the partial methods
2. Add a rule that fails all aggregates that don't support partial aggregation for now.

The first option is the easiest, but it means that this API is less expressive. As @cloud-fan noted, partial aggregation doesn't make sense for all aggregate functions.

Reply:
> I don't think Spark can change this in the near future

Have we (i.e., Spark community) thought about removing this restriction in future? If there's no motivation to do this then IMO option 1) is better since otherwise having the `AggregateFunction` is a little confusing as ppl will always use `PartialAggregateFunction`. Otherwise, I'd go with option 2).

With option 2) we can just simply reject non `PartialAggregateFunction` s right after they are loaded from the catalog right?

Reply:
I'd probably prefer the first option if there aren't many opinions out there about this. It is disappointing that Spark doesn't support simpler aggregates, though.

Reply:
It's very hard to support this, as one aggregate operator may host many aggregate functions that some support partial merging but some do not. I'm +1 to option 1.